### DISCUSSION TITLE
Assembling to Scalar Variable Residual from Mortar Constraint

### DISCUSSION THREAD

**Discussion by ttruster** [idaholab/moose#22020] [Q&A General] [answered]
I find myself needing to implement coupling between a scalar variable into a mortar constraint residual and Jacobian and vice versa. I have looked closely over the ScalarLagrangeMultiplier Kernel and the AverageValueConstraint Scalar Kernel for diffusion as well as Homogenization files (Constraint user object, scalar kernel, and off-diagonal Jacobian kernel) for the Tensor Mechanics system. It seems that the "MOOSE way" to do this would require a:
1. Mortar Constraint for the off-diagonal Jacobian (I have implemented)
2. Scalar Kernel to handle the residual and Jacobian of the scalar variable (I have made a hard-coded version)
3. A User Object that performs integration over a Mortar lower dimensional boundary object to compute the terms needed in the Scalar Kernel (non-existent in the MOOSE framework or modules; would require some framework expansion)

As a temporary work-around, I have tried adding direct assembly calls within the Mortar Constraint object for vector-residual and matrix-Jacobian blocks associated with the scalar variable. I'm using calls similar to those in ScalarLagrangeMultiplier and HomogenizedTotalLagrangianStressDivergence, which assembly to the upper and lower blocks of the Jacobian at the same time. They are happening through additional calls (quadrature loops and functions) that I embedded within the ComputeResidual and ComputeOffDiagonalJacobianScalar method calls of the MortarConstraint and MortarConstraintBase files. I have attached a sample of the source code with the loops.
[src.txt](https://github.com/idaholab/moose/files/9500175/src.txt)

I can confirm from debug-mode that computations are occurring in the Residual Scalar calculation during the time when the Mortar Constraint ComputeResidual method is called. However, it doesn't seem like those values are getting into the global system residual, since the analysis finishes with a zero Nonlinear Residual norm.

So, question:  am wondering if it is not possible to assemble residual contributions into other/coupled variables (i.e. the test function rows) different from the primary variable of the object, without modifying the tagging/caching/low-level assembly calls within MOOSE framework that are looping over the objects? The symmetric-block access for the Jacobian within ScalarLagrangeMultiplier gave me hope that would be possible. Of course it is not "appropriate" according to MOOSE design, but didn't know if it is technically not feasible.

The long term resolution is to make a user-object that has reference to the mortar subdomain mesh and system, to perform the quadrature. I might be able to get that to happen through either a SideUserObject.execute or a DomainUserObject.executeOnBoundary call, if I can get the Two Material System coupled into those. The MortarFrictionalPressureVectorAux AuxKernel gives me hope for doing that, since it also references the mortar mesh and data structure but outside of a Mortar Constraint object. Any clues for going down this route?

Thanks for reading, and suggestions welcome!

Tim

**Comment by reverendbedford** [comment]
This is exactly analogous to the [homogenization constraint](https://mooseframework.inl.gov/docs/doxygen/modules/classHomogenizationConstraintIntegral.html) stuff in the Lagrangian kernels, right?  So I take it you can't use a `ElementUserObject` to loop over the lower-dimension mortar blocks?  Shouldn't they just be standard 2D domains (in a 3D problem)?

**Comment by lindsayad** [comment]
The problem is that the "local" scalar residual information is not getting incorporated into global data. Here are revised code snippets that should work
```c++
void
VMNTPeriodicBase::::computeResidual()
{
  setNormals();

  precalculateResidual();

  if (_compute_primal_residuals)
  {
    // Compute the residual for the secondary interior primal dofs
    computeResidual(Moose::MortarType::Secondary);

    // Compute the residual for the primary interior primal dofs.
    computeResidual(Moose::MortarType::Primary);
  }

  if (_compute_lm_residuals)
    // Compute the residual for the lower dimensional LM dofs (if we even have an LM variable)
    computeResidual(Moose::MortarType::Lower);

  if (_compute_scalar_residuals)
    // Compute the residual for the scalar dofs
    computeResidualScalar();
}



void
VMNTPeriodicBase::computeResidualScalar()
{
  prepareVectorTag(_assembly, _kappa_var);
  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
  {
    precalculateMaterial();
    for (_h = 0; _h < _k_order; _h++)
      _local_re(_h) += _JxW_msm[_qp] * _coord[_qp] * computeQpResidualScalar();
  }

  accumulateTaggedLocalResidual();

  for (const auto tag_id : _vector_tags)
  {
    const auto & vector_tag = _subproblem.getVectorTag(tag_id);
    _assembly.addResidualScalar(vector_tag);
  }
}
```
The call to `_assembly.addResidualScalar` is the important new piece that takes the local residual and adds it into global residual information

**Comment by ttruster** [comment]
Thanks for suggestions. I've got a summary at the bottom, about the topic of add-Jacobian needs and about user objects.

I switched the residual assembly to use the tagging system. The compiler complained that 
`Assembly::addResidualScalar(const VectorTag&)’ is protected within this context`
And so instead I made a call to 
`_fe_problem.addResidualScalar();`
That is in the attached file, with all the other assemblies I am trying to run.
[VMNTPeriodicBase.txt](https://github.com/idaholab/moose/files/9530420/VMNTPeriodicBase.txt)

I was able to run a test case for heat conduction with mortar periodic conditions and a macro heat flux (analogous to macro stress) imposed, and the analysis solves for the macro thermal gradient as a scalar variable and the temperature field as a Lagrange variable. The analysis converges with the PJFNK solver and gives the correct result for both solution fields.

However, now this issue is that the off-diagonal mortar constraint Jacobian (coupling secondary variable and scalar variable) seem not to be accumulating/adding. When I use the NEWTON solve type, then the residual diverges. When I comment out or otherwise change around the code in the "computeOffDiagJacobianScalar" function, the residual normal are identical, which suggests to me that this code isn't getting added to the global Jacobian.

I've been self-studying on the Problem, System, Loops, and Assembly source files through doxygen to try and learn the ways that local residual/Jacobian blocks get assembled to the global arrays. It seems that each object type (kernel, dgkernel, scalarkernel, ...) get handled at different places; some of them involve caching while the "ComputeMortarFunctor" file seems to have less provisions for scalar and off-diagonal variable couplings.

I've made a couple attempts at getting the local data into the global; those are in the comments of the attached .txt (a .C) file.
One thing that mystifies me: the "ScalarLagrangeMultiplier" object only contains two lines that give an allocation to a DenseMatrix; there are no calls to an accumulate or an add-Jacobian function below that. My gut says that probably those values need to go into the specially named ken and kne arrays, and that they are swept up by the caching that happens in "ComputeFullJacobianThread" and "NonlinearSystemBase". But without knowing where to look, I'm not sure what code to copy in order to handle the Secondary-scalar, Primary-Scalar, and Lower-Scalar (and the transposes) assemblies that I need to accumulate.

In short, I've been gathering a list of needed assembly and pre-calculate Residual etc. functions that are available for Kernels and are missing in some of the object systems. If it helps, I could be working on the framework to add those and then do a pull request; this would be my first time, so some guidance would help (i.e. I probably need to read this https://mooseframework.inl.gov/framework/contributing.html)

About User Objects: yeah I've been thinking about the minimal integrals needed and I may be able to get by with two single-side integrals rather than having to do a coupled Mortar integral, if I can somehow grab material properties from the opposite element face across the mortar segment pair. It's yet to be seen whether I should try that with an Element user object on 2d blocks (can they be called on the mortar sub-problem?) or if instead needs to happen on the 3d problem Side sets user object.

**Reply by lindsayad** [reply]
The problem with

> _fe_problem.addResidualScalar();

is that it's not going to honor the vector tags requested by the user for their constraint object. It's going to add to all vectors that are of residual type. A better solution is to make the individual vector tag API that is currently `protected` `public`.

**Reply by lindsayad** [reply]
> About User Objects: yeah I've been thinking about the minimal integrals needed and I may be able to get by with two single-side integrals rather than having to do a coupled Mortar integral, if I can somehow grab material properties from the opposite element face across the mortar segment pair. It's yet to be seen whether I should try that with an Element user object on 2d blocks (can they be called on the mortar sub-problem?) or if instead needs to happen on the 3d problem Side sets user object.

For fundamental assembly operations I would be hesitant about a design that includes user objects. If user objects are being included, to me that indicates that we don't have friendly enough APIs on our `_sys`, `_assembly`, or `_subproblem` to get your job done, when we really should.

**Reply by lindsayad** [reply]
There is definitely some help we need to give you here and at least some framework development that will need to happen. I need to think about this a bit more in order to give you good help (hopefully tomorrow). In the mean time, if you are motivated to do some moose development, please do! We love contributions!

**Reply by hugary1995** [reply]
Regarding framework support, I wonder if we can't add an "integrated" scalar kernel which loops over the elements to assemble the residual, Jacobian, and off-diagonal Jacobians for a scalar variable. The objects Tim mentioned, e.g. ScalarLagrangeMultiplier, AverageValueConstraint, and HomogenizationConstraintIntegral, will all benefit from having such loop in the framework. 

The main reason that the Homogenization system in tensor mechanics need to use an ElementUserObject is to retrieve material properties. If we embrace that integrated scalar kernel idea, then we will no longer need such a user object, since we can bring the MaterialInterface into the integrated scalar kernel.

**Reply by ttruster** [reply]
Having "integrated" scalar kernels to allow coupling of regular variables and scalar variables would definitely be one way to solve the problems above. They originate from needing to evaluate weak form terms (integrals) have products of variables, scalars, and materials.

The "issue" is that those integrals would be needed (in the long term) over all the different types of integrals. For my current project goals, I will end up needing integrals for kernels, interfacekernels, and mortar constraint objects to couple with scalar variables. DG-kernels and integrated BCs would be later in the future if cohesive-zone-like damage would be included.

So the current approach with the ScalarLagrangeMultiplier, AverageValueConstraint, and HomogenizationConstraintIntegral objects has been instead to use the integration methods that are present within user objects to handle the regular variable and material couplings, and then multiply the integral result onto scalar variable within the kernel; and just make sure the user object gets called on each linear execute.

So that seems as a design philosophy question, as to whether enough folks see the utility to have the full coupling ability and put all the scalar integrals as derivatives off of the scalar kernel. Or to just finish expanding the coverage of user objects to handle mortar constraint objects or any other "missing" cases. 

That's where my hiccups have been so far, is needing to figure out which assembly accumulation is missing for various coupling, as well as how to get the integrals done.

I figured my hack for calling the _fe_problem.addResidualScalar(); might not be great, but I figure that the tagging API is probably protected so that most normal users/developers don't "mess up" or call in the tag additions where they shouldn't be.

**Comment by ttruster** [comment]
Well, I spent a few hours on Friday night and then did some thinking over the last few days related to the topic of mortar user objects and assembly of scalar variable contributions. I'm going to break the assembly off into another thread to raise an issue there.

For the Mortar User Object, I'm pretty sure what needs to happen is to add this object class into `computeUserObjectsInternal` as well as the initialization locations for user objects, so that this is on the same level as element, side, interface, or domain user objects. Thus, a loop over all mortar user objects would be needed, and so I should emulate what `computeMortarFunctor` does for looping over `MortarConstraintBase` objects to collect the residual and/or Jacobian contributions.

Adding this new type of user object is necessary because:

1. Only specialized mortar objects are able to generate the container of material properties and other data needed for the mortar segment calculations, namely `loopOverMortarSegments()`
2. Even for conforming primary/secondary meshes in 3d, for history-dependent or complex material models, the use of quadrature points from a full element side integral are different than the segments, so the accuracy of the computations would degrade compared to using the same quadrature points for both user object and constraint residual integrals

I would like to open an issue with the objective of programming a "MortarUserObjectBase" class and the necessary initialization/loops to fit within the user object system. The while the base class would not contain integrals (so you could make other types of mortar objects from it), a derived class with the integration would also be developed.

The target application of these developments are:
Problems involving scalar variables coupled to standard variables along mortar surfaces. An example of this class of problems is: imposition of periodic boundary conditions through surface integration and either penalty, Nitsche, or Lagrange multiplier field methods.

I've been spending time looking through the framework code, and I have some ideas to get started with, I'll need some discussion/help in the process. This will be my first time working on an issue; I've already forked the main repository and am starting to make commits on the side.

**Reply by lindsayad** [reply]
Is there more to this problem than just missing Jacobian accumulations for the scalar variable? That seems like just one missing API call. We are going down what seems like too large a development path if that is the only problem. The missing Jacobian accumulation call is likely very akin to what was the missing residual accumulation call. If that is the only problem, then if you create a branch with a MNWE (minimal non-working example), then I think we could get this solved pretty quickly. (I generally prefer branches over shared text files because then it is much easier to document improvements via commits)

**Comment by ttruster** [comment]
We can avoid the user object route if we do 3 of the 4 needed Jacobian (scalar to variable, variable to scalar, and scalar to scalar) and 2 of the 2 needed residual (variable and scalar) computations all within one Mortar Constraint. This was the "hacky" way I proposed at the top, since it seems to violate the MOOSE goal of 'each object compute and assemble contributions only to the variable it owns', i.e. to one row of the Jacobian and residual (variable to variable handled in another constraint file, and scalar to variable). I'm also wondering if there is potential value to open up user objects and post processors that do loops over mortar segments for some other purposes.

I terms of a MNWE, I have my forked repo here: https://github.com/Truster-Research-Group/moose-clmi_lab
Again, first timer: should I put the test input file in the /test/test directory somewhere, the needed .C and .h files for the code I've created which needs the accumulate commands within /test/src and /test/include respectively, and make a commit saying what file to run, push that commit to my repo, and let you know about it by posting on the forum here? Or can you paste a link to a forum post that has the process I should emulate?

**Reply by lindsayad** [reply]
What is this scalar variable? What does it represent physically? Does it have other residual objects that contribute to its residual/Jacobian? I would say that the MOOSE philosophy is to (loosely) have an object per physics/PDE-term. Mortar constraints themselves can already contribute residuals/Jacobians to three different variables (secondary variable, primary variable, LM variable).

> Again, first timer: should I put the test input file in the /test/test directory somewhere, the needed .C and .h files for the code I've created which needs the accumulate commands within /test/src and /test/include respectively, and make a commit saying what file to run, push that commit to my repo, and let you know about it by posting on the forum here?

I would just keep your code in whatever directories you already have them in. You can also put your input wherever you want. I can see where all your files are located just by running commands like `git show --name-only HEAD`. Before committing and pushing you might want to checkout a new branch (something like `git checkout -b non-working-scalar-in-mortar`)

**Comment by ttruster** [comment]
Hit the enter key...

There are two scalar variables for this problem: the macro flux-gradient prescribed through an aux-variable, and the macro actual gradient solved for alongside the concentration field. For solid mechanics, these are analogous to macro-stress and macro-strain, respectively, applied to an RVE. Theory for Nitsche method version is described here https://www.sciencedirect.com/science/article/pii/S0045782519305481 and the mortar version is described here https://www.sciencedirect.com/science/article/pii/S0045782514000528.

The scalar variable is coupled/integrated along each RVE boundary pair, which likely are doing to be different mortar surfaces. Yet they are pretty much going to involve the same variables each time. 
As seen in the example, the residual/Jacobian terms of the primary to secondary variable can be handled in a different object than the terms involving products of the primary/secondary with the scalar variable. Or you could put those terms within the same object with the scalar; I split them up here to illustrate the point.
In short, the term from the paper above that I've implemented is { [w1-w2] - dx*kappa } * tau * { [u1-u2] - dx*epsilon } where 1=primary, 2=secondary, epsilon is the scalar unknown (macro gradient [u,x u,y]), u is the variable unknown, dx = [x1-x2, y1-y2] in 2d, and w/kappa are the weighting functions. So if you call this "one term", then yes it can all go into one object.


About the performance: right now, the PJFNK method is used so that it will converge. Using Newton doesn't converge because the off diagonal Jacobian Scalar terms aren't getting assembled. I have also left the residual assembly with the `_fe_problem` object instead of `_assembly.addResidualScalar` for the time being. But I'm going to be adding some other examples in the next couple days that try that after changing it to a public method.

Let me know your thoughts, thanks.

**Reply by lindsayad** [reply]
Is the branch name `non-working-scalar-in-mortar`?

**Reply by lindsayad** [reply]
@jbadger95, Nitsche method (but for periodic boundary constraints as opposed to mechanical contact constraints)

**Reply by ttruster** [reply]
Yes, that's the branch; others are for other in progress work.

**Comment by ttruster** [comment, marked as answer]
I added another test case of scalar variables using the Kernel this time, into the `non-working-scalar-in-mortar`. There is a base class `KernelScalarBase` which expands the Kernel computations to act on a second coupled variable (a scalar). I created it using `LowerDIntegratedBC` as a model for how to add the contributions from the 2nd variable and reuse the quadrature loops and methods of the underlying class.

As a test case, I copied over the `ScalarLagrangeMultiplier` and `AverageValueConstraint` contributions to be in the single derived class `ScalarLMKernel`. I also copied the test case `scalar_constraint_kernel.i` and modified it to have syntax for this new object.

There are some comments in the file about 'temporary' fixes to make it execute. The good news is, the solution converges to the same analytical result that the existing 3 object system (with a post processor integral) does, which means the residual is being assembled; the bad news is that the convergence is slow, so something is off with the Jacobian. I literally copied the Dense Matrix calls from the existing objects, and this is deriving from the same Kernel class; although I am using the Kernel's off diagonal Scalar method for computing one of the contributions.

Thoughts? Again, some Jacobian accumulation/adding seems to be missing.

**Reply by ttruster** [reply]
I spoke too soon and didn't look closely at the iteration type: in fact, it is doing linear/matrix solver iterations, not Newton iterations.

I confirmed after adding an off diagonal scalar method using dense matrix assembly, identical to `ScalarLagrangeMultiplier`, that the linear iterations are identical for the new and old systems. Also, with a new input file `scalar_constraint_kernel_LU.i` which uses an LU solver and Newton method, that indeed there is just one iteration.

So the Dense Matrix approach seems to work for the Kernel objects, but does not work for the Constraint or Interface-Kernel objects (yet). Thoughts on why that is?

Also, I've been wondering how either the Dense Matrix or the tagged scalar vectors/matrices will work in parallel (threaded/OpenMP or distributed/MPI) cases? I've been running it with one thread so far. The adding of the residual from multiple threads to the same scalar variable rows probably creates a race condition.

**Reply by lindsayad** [reply]
we have mutex locks to ensure no race conditions when attempting to add or insert into the global vector or matrix

**Reply by lindsayad** [reply]
> So the Dense Matrix approach seems to work for the Kernel objects, but does not work for the Constraint or Interface-Kernel objects (yet). Thoughts on why that is?

Different calls to different assembly methods after the residual object jacobian routines. I hope to run your code soon to verify. Fixing this should just be a couple line addition in framework code

**Reply by ttruster** [reply]
Thanks. I also made a couple more commits related to the kernel plus scalar coupling. Now that class can act as a pseudo-intermediate class to add coupling to scalars but not require the scalar to be used in the derived classes. This would help for the tensor mechanics homogenization case, where that class derives 3 levels down from the Kernel class. The example test shows the Diffusion Kernel can derive from it instead and not use a scalar.
Maybe these can point out some things about assembly too.

**Reply by lindsayad** [reply]
There are so many routines in `TestPeriodicBase` it's hard for me to wrap my head around what is adding to what. Can you tell me exactly what routines do these couplings?
- field-scalar
- scalar-field

I assume field-scalar is done solely by `computeOffDiagJacobianScalar`, but I want to be sure.

Please respond within this thread (via 'Reply') while we're on this topic. It's hard for me to keep track of the thread of conversation when there are multiple Answers addressing the same thread of conversation.

**Reply by lindsayad** [reply]
Actually I've created issue #22174, so let's continue this over there.
